### ISSUE TITLE
data layout

### ISSUE THREAD

**Issue by ThisIsIsaac** [facebookresearch/flashlight#58]
Arrayfire's data layout is WHCN, so I was assuming flashlight would have the same layout but I found some code like this in [CudnnUtils.cpp](https://github.com/facebookresearch/flashlight/blob/master/flashlight/autograd/backend/cuda/CudnnUtils.cpp)

```cpp
  // reverse the arrays and cast to int type
  std::array<int, 4> strides = {(int)afstrides[3],
                                (int)afstrides[2],
                                (int)afstrides[1],
                                (int)afstrides[0]};
```

1. What is the data layout of flashlight?
2. why do you reverse the layout at the last moment before calling cuDNN?
3. does that mean you also flip padding, stride, and dilation?

**Comment by vineelpratap** [comment]
Hi,
 Arrayfire uses column major layout. So, `WHCN` layout in arrayfire is same as cuDNN NCHW layout once you get the data pointer for the array. Hope that answers the remaining questions.

**Comment by ThisIsIsaac** [comment]
@vineelpratap

A couple of followup questions:

1. what do you mean by **column major layout**? I am used to thinking of memory as a one dimensional array on the hardware and column-major and row-major is just how we chose to interpret the 1D array. So when we are talking about how elements of a tensor are laid out on the memory, I don't understand what you mean by **row** or **column**.

2. I am trying to read an array from Pytorch into flashlight. Pytorch and cuDNN uses NCHW layout where elements along W dimension are stored consecutively, then you have consecutive rows of W stored along the H dimension, and so on. How would I change the data layout of a Pytorch array to correctly read it into an array fire array?

**Comment by ThisIsIsaac** [comment]
Array fire is **HWCN, not WHCN**. I had problems porting layers from Pytorch to Flashlight because we thought Arrayfire was WHCN.

In the comments ([source](https://github.com/facebookresearch/flashlight/blob/master/flashlight/autograd/backend/cpu/BatchNorm.cpp#L69)):

> tensor is in WHCN order and layout in COLUMN-MAJOR

However, this is not true. Although within flashlight WHCN is equivalent to HWCN as long as you keep it consistent, this becomes a problem when we need to port from another framework such as Pytorch or Tensorflow into Flashlight. We should fix the comment

**Comment by vineelpratap** [comment]
> - row-major NCHW is equivalent to column-major WHCN in terms of data layout
> - one can just simply copy weights from Pytorch's tensor into Flashlight without rearranging the dimensions.

@ThisIsIsaac  - yes, these are both true. 



> Array fire is HWCN, not WHCN. I had problems porting layers from Pytorch to Flashlight because we thought Arrayfire was WHCN.

I think the original comment is correct - input is WHCN for column-major layout. This actually follows your analysis from above two statements -->
`one can simply reuse the Pytorch's input tensor in NCHW format for convolution` 

Can you give an example for this, otherwise ?

**Comment by ThisIsIsaac** [comment]
I've been asking for clarifications from AF dev via slack, and this is one of his responses:

> Say an image of size WxH and C channels is loaded by ArrayFire, the data on arrayfire's array would be of size [ H W C 1 ] , scanlines sit as rows in ArrayFire array 

If you want, I am going to ask my junior dev -- @minty99 -- about a specific example led us to thinking that AF is HWCN

**Comment by tlikhomanenko** [comment]
@ThisIsIsaac 

Check this example http://arrayfire.org/docs/machine_learning_2kmeans_8cpp-example.htm#a13, seems even `loadImage` function loads image in `WHC` format.
```
array img = loadImage(ASSETS_DIR"/examples/images/vegetable-woman.jpg", true) / 255; // [0-255]
int w = img.dims(0), h = img.dims(1), c = img.dims(2);
```

**Comment by minty99** [comment]
My reason to say arrayfire is HWCN is this -

```
af::array a = af::randn(2, 3);
af::array b = af::randn(2, 5);
af::print("2x3", a);
af::print("2x5", b);
```

prints

```
2x3
[2 3 1 1]
   -0.9247     2.5441    -0.3452 
    0.1808     0.3516     0.2191 

2x5
[2 5 1 1]
   -0.7687    -1.1948    -0.5378     0.9354     0.5787 
    0.2413     0.8927     0.2270    -0.7613    -0.6174 
```

If arrayfire is WHCN, this test should print 3-by-2 matrix and 5-by-2 matrix.

**Comment by vineelpratap** [comment]
@minty99 - We shouldn't use print output to conclude anything about the data layout, since it doesn't show how the underlying raw data is stored in memory. You can try using `.host` operator to get the raw pointer for the array. 

As @tlikhomanenko mentioned, you can try to load a rectangular image and see the dimension of the output. Also, note that height and width makes sense only when you are operating on images, while convolutions can be applies on any other domain. 

> one can just simply copy weights from Pytorch's tensor into Flashlight without rearranging the dimensions.

Based on this statement above, you can also try using the same input, weight in PyTorch 2D convolution and it should produce the same output. This should also confirm that flashlight using WHCN.

**Comment by ThisIsIsaac** [comment]
@vineelpratap  Thanks. We are going to copy weights from Pytorch and see if it actually works.
